### PR TITLE
Improve ST construction & ser/des interfaces:

### PR DESCRIPTION
**Pull request by nbougalis**
## High Level Overview of Change

The primary motivation of this commit is keep the serialization and deserialization interfaces separate and, hopefully, provide cleaner, simpler and less error-prone interfaces.

The deserializer is now also as close to zero-copy as possible, especially for "VL" fields (aka blobs).

Multiple ST* constructors have been consolidated, allowing for more uniform and readable construction of objects, and reducing code duplication and unnecessary copying of data.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

File: src/ripple/rpc/handlers/Tx.cpp [hunk 1]
```diff
@@ -119,9 +119,9 @@ doTxPostgres(RPC::Context& context, TxArgs const& args)
             if (args.binary)
             {
                 SerialIter it(item->slice());
-                it.skip(it.getVLDataLength());  // skip transaction
-                Blob blob = it.getVL();
-                res.meta = std::move(blob);
+                (void)it.getVL();  // skip the transaction
+                Slice s = it.getVL();
+                res.meta = std::move(Blob{s.begin(), s.end()});
```
Comment (HowardHinnant) [inline src/ripple/rpc/handlers/Tx.cpp:124]:
```
src/ripple/rpc/handlers/Tx.cpp:124:28: warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
                res.meta = std::move(Blob{s.begin(), s.end()});
                           ^
src/ripple/rpc/handlers/Tx.cpp:124:28: note: remove std::move call here
                res.meta = std::move(Blob{s.begin(), s.end()});
                           ^~~~~~~~~~                        ~
```

File: src/ripple/shamap/impl/SHAMapTreeNode.cpp [hunk 2]
```diff
@@ -86,24 +89,24 @@ SHAMapTreeNode::makeAccountState(
...
-    // FIXME: improve this interface so that the above check isn't needed
-    if (!s.getBitString(tag, s.size() - tag.bytes))
-        Throw<std::out_of_range>(
-            "Short AS node (" + std::to_string(s.size()) + ")");
-
-    s.chop(tag.bytes);
-
-    if (tag.isZero())
+    auto const tag = [&data]() {
+        std::optional<uint256> ret;
```
Comment (HowardHinnant) [inline src/ripple/shamap/impl/SHAMapTreeNode.cpp:93]:
```
src/ripple/shamap/impl/SHAMapTreeNode.cpp:93:32: warning: unused variable 'ret' [-Wunused-variable]
        std::optional<uint256> ret;
                               ^
```

**Review by HowardHinnant — APPROVED** [review]

**Comment by intelliot** [comment]
@nbougalis if this PR should still be considered for merging, please resolve conflicts and bring it up-to-date with `develop`, preferably with a merge commit. Thanks!

**Comment by nbougalis** [comment]
I think the PR should be merged, but frankly rebasing it now is a bit of a pain, since another commit went ahead of it, reorganizing the code significantly. I'm not sure when I'll be able to rebase this.

**Comment by nbougalis** [comment]
Rebased. It should be re-reviewed by @HowardHinnant and hopefully one other person, because the merge was mildly complicated. Once it gets a sign-off, this should be good to go.

**Pull request closed without merging**